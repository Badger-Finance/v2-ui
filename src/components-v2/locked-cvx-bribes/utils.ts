import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
