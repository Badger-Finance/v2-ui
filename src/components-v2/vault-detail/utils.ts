import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
