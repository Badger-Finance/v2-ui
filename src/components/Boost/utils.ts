import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
