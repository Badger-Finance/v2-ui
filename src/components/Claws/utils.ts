import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
