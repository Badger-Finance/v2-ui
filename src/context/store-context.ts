import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;