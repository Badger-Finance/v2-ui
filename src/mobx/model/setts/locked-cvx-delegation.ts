export enum DelegationState {
	Eligible = 'eligible',
	Ineligible = 'ineligible',
	Delegated = 'delegated',
	BadgerDelegated = 'badger-delegated',
}
