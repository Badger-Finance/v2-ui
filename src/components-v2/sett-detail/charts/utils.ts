import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
