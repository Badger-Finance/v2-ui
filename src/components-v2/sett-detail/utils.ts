import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
