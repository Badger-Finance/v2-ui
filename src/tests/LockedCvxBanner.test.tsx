import React from 'react';
import store from '../mobx/RootStore';
import { customRender, fireEvent, screen } from './Utils';
import { StoreProvider } from '../mobx/store-context';
import LockeDelegationBanner from '../../src/components-v2/locked-cvx-bribes/Banner';
import { checkSnapshot } from './utils/snapshots';
import { DelegationState } from '../mobx/model/setts/locked-cvx-delegation';
import BigNumber from 'bignumber.js';
import LockedCvxDelegationStore from '../mobx/stores/lockedCvxDelegationStore';

describe('Locked CVX Delegation Banner', () => {
  beforeEach(() => {
    store.lockedCvxDelegation.delegationState = DelegationState.Eligible;
    store.lockedCvxDelegation.totalEarned = new BigNumber(100 * 1e18);
    store.lockedCvxDelegation.lockedCVXBalance = new BigNumber(100 * 1e18);
    store.lockedCvxDelegation.unclaimedBalance = new BigNumber(100 * 1e18);

    jest.spyOn(LockedCvxDelegationStore.prototype, 'canUserDelegate', 'get').mockReturnValue(true);
    jest.spyOn(LockedCvxDelegationStore.prototype, 'shouldBannerBeDisplayed', 'get').mockReturnValue(true);
  });

  it('renders correctly', () => {
    checkSnapshot(<LockeDelegationBanner />);
  });

  it('can click delegate locked cvx', async () => {
    const mockDelegateLocked = jest.fn();
    store.lockedCvxDelegation.delegateLockedCVX = mockDelegateLocked;

    customRender(
      <StoreProvider value={store}>
        <LockeDelegationBanner />
      </StoreProvider>,
    );

    fireEvent.click(screen.getByRole('button', { name: 'Delegate to Badger' }));

    expect(mockDelegateLocked).toHaveBeenCalled();
  });
});
