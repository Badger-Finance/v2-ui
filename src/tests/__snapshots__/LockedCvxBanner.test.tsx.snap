// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locked CVX Delegation Banner renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root-12 MuiCard-root-11 makeStyles-root-1 MuiPaper-elevation1-16 MuiPaper-rounded-13"
  >
    <p
      class="MuiTypography-root-40 makeStyles-title-3 MuiTypography-body1-42"
    >
      Delegate to Earn Rewards
    </p>
    <div
      class="makeStyles-blueParticle-7"
    />
    <div
      class="makeStyles-greenParticle-8"
    />
    <img
      alt="diagonal-particles"
      class="makeStyles-diagonalParticles-9"
      src="assets/diagonal-particles.svg"
    />
    <div
      class="MuiGrid-root-70 MuiGrid-container-71 MuiGrid-spacing-xs-3-95"
    >
      <div
        class="MuiGrid-root-70 MuiGrid-container-71 MuiGrid-item-72 MuiGrid-grid-xs-12-116 MuiGrid-grid-sm-true-118"
      >
        <img
          alt="cvx locked safebox illustration"
          class="makeStyles-safeLockImageSm-6"
          src="assets/locked-cvx-safebox-sm.png"
        />
        <div
          class="makeStyles-descriptionSection-4"
        >
          <p
            class="MuiTypography-root-40 makeStyles-descriptionText-10 MuiTypography-body1-42"
          >
            Earn a portion of the Votium incentives by simply delegating your Vote Locked Convex to Badger today.
          </p>
        </div>
        <div
          class="MuiPaper-root-12 MuiCard-root-11 MuiGrid-root-70 makeStyles-delegationSection-2 MuiGrid-container-71 MuiPaper-elevation1-16 MuiPaper-rounded-13"
        >
          <div
            class="MuiGrid-root-70 MuiGrid-container-71 MuiGrid-spacing-xs-2-94 MuiGrid-align-items-xs-center-79"
          >
            <div
              class="MuiGrid-root-70 makeStyles-section-176 MuiGrid-item-72 MuiGrid-grid-xs-12-116 MuiGrid-grid-md-true-132"
            >
              <p
                class="MuiTypography-root-40 makeStyles-delegationTitle-177 MuiTypography-body1-42"
              >
                Vote Locked Convex
              </p>
              <div
                class="MuiGrid-root-70 MuiGrid-container-71 MuiGrid-align-items-xs-center-79 MuiGrid-justify-content-xs-center-88"
              >
                <img
                  alt="locked cvx balance"
                  class="makeStyles-icon-173"
                  src="assets/icons/bvecvx.png"
                />
                <h4
                  class="MuiTypography-root-40 makeStyles-delegationAmount-174 MuiTypography-h4-48 MuiTypography-displayInline-68"
                >
                  100
                </h4>
              </div>
            </div>
            <div
              class="MuiGrid-root-70 makeStyles-section-176 MuiGrid-item-72 MuiGrid-grid-xs-12-116 MuiGrid-grid-md-true-132"
            >
              <button
                class="MuiButtonBase-root-209 MuiButton-root-180 MuiButton-contained-188 MuiButton-containedPrimary-189 MuiButton-containedSizeSmall-199 MuiButton-sizeSmall-201"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label-181"
                >
                  Delegate to Badger
                </span>
                <span
                  class="MuiTouchRipple-root-220"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root-70 MuiGrid-item-72 MuiGrid-grid-xs-12-116 MuiGrid-grid-sm-true-118"
      >
        <div
          class="MuiPaper-root-12 MuiCard-root-11 makeStyles-root-212 MuiPaper-elevation1-16 MuiPaper-rounded-13"
        >
          <div
            class="MuiGrid-root-70 makeStyles-container-213 MuiGrid-container-71"
          >
            <p
              class="MuiTypography-root-40 makeStyles-earningsTitle-219 MuiTypography-body1-42"
            >
              Earned Incentives
            </p>
            <div
              class="MuiGrid-root-70 makeStyles-statRow-214 MuiGrid-container-71"
            >
              <h6
                class="MuiTypography-root-40 makeStyles-statName-215 MuiTypography-subtitle1-51"
              >
                Total Claimed
              </h6>
              <div
                class="MuiGrid-root-70 MuiGrid-container-71 MuiGrid-align-items-xs-center-79"
              >
                <img
                  alt="badger coin image"
                  class="makeStyles-badgerImage-218"
                  src="assets/icons/badger.png"
                />
                <h5
                  class="MuiTypography-root-40 makeStyles-stat-216 MuiTypography-h5-49 MuiTypography-displayInline-68"
                >
                  100
                </h5>
              </div>
            </div>
            <div
              class="MuiGrid-root-70 makeStyles-statRow-214 MuiGrid-container-71"
            >
              <h6
                class="MuiTypography-root-40 makeStyles-statName-215 MuiTypography-subtitle1-51"
              >
                Unclaimed
              </h6>
              <div
                class="MuiGrid-root-70 MuiGrid-container-71 MuiGrid-align-items-xs-center-79"
              >
                <img
                  alt="badger coin image"
                  class="makeStyles-badgerImage-218"
                  src="assets/icons/badger.png"
                />
                <h5
                  class="MuiTypography-root-40 makeStyles-stat-216 makeStyles-unclaimedAmount-217 MuiTypography-h5-49 MuiTypography-colorPrimary-63 MuiTypography-displayInline-68"
                >
                  100
                </h5>
                <button
                  class="MuiButtonBase-root-209 MuiButton-root-180 MuiButton-contained-188 MuiButton-containedPrimary-189"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label-181"
                  >
                    Claim Bonus
                  </span>
                  <span
                    class="MuiTouchRipple-root-220"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <img
      alt="safe-lock-image"
      class="makeStyles-safeLockImageLg-5"
      src="assets/locked-cvx-safebox-lg.png"
    />
  </div>
</div>
`;
